### PR TITLE
HDDS-7930. input stream does not refresh expired block token.

### PR DESCRIPTION
**Pull request by jojochuang**
## What changes were proposed in this pull request?

Update the retry function to not only update pipeline, but also updating block token.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7930

## How was this patch tested?
Unit tests

### CODE REVIEW

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 1]
```diff
@@ -199,17 +199,19 @@ private boolean isConnectivityIssue(IOException ex) {
...
+    if (refreshFunction != null) {
+      LOG.debug("Re-fetching pipeline and block token for block {}", blockID);
+      BlockLocationInfo blockLocationInfo = refreshFunction.apply(blockID);
+      if (blockLocationInfo == null) {
+        LOG.debug("No new block location info for block {}", blockID);
       } else {
-        LOG.debug("New pipeline for block {}: {}", blockID, newPipeline);
-        this.pipeline = newPipeline;
+        LOG.debug("New block location info for block {}: {}",
+            blockID, blockLocationInfo);
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:212]: `BlockLocationInfo#toString` includes the token.  I think we should keep logging the pipeline only.

**Review by adoroszlai — COMMENTED** [review]
LGTM, with minor change suggested for logging.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by kerneltime** [comment]
Refresh pipeline is used on failover, once the improved block token design is in the performance of refresh of block tokens won't be an issue. I guess for the short time this is ok to refresh block tokens along with pipeline.

**Comment by duongkame** [comment]
> Refresh pipeline is used on failover, once the improved block token design is in the performance of refresh of block tokens won't be an issue. I guess for the short time this is ok to refresh block tokens along with pipeline.

The refresh function calls getKeyInfo and OM generates a new block token every refresh already. Before this change, the client input stream doesn't take the new block token in the result. (And that sounds like a bug to me)

**Review by duongkame — APPROVED** [review]
LGTM.

**Comment by jojochuang** [comment]
Merged. Thanks all @szetszwo @duongkame @kerneltime @adoroszlai

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java (hunk 1 above)
Comment (jojochuang) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:212]: omg i missed this one. Will add an addendum to it.

**Pull request merged**